healthcheck/healthcheck.go

```diff
@@ -0,0 +1,157 @@
+	"sync"
+	"sync/atomic"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const (
+	livenessPath = "/liveness"
+	readinessPath = "/readiness"
+	portNum = ":8080" // TODO(monazhn): Think about a good port number.
+)
+
+// Locks to synchronize reads and writes for each HC boolean. Separate locks
+// for each boolean is ideal so more than one boolean can be read/written
+// simultaneously.
+var (
+	readinessL = &sync.Mutex{}
+	livenessL  = &sync.Mutex{}
+	startupL   = &sync.Mutex{}
+)
+
+// HC is a type used to implement health checks for the proxy.
+type HC struct {
+	// live being true means the proxy is running; in the case of the proxy
+	// being unexpectedly terminated, we should (re)start the proxy.
+	// live is related to Kubernetes liveness probing.
+	live  bool
+	// ready being true means the proxy is ready to serve new traffic; in the 
+	// case that ready is false, we should wait to send new traffic to the
+	// proxy. The value of ready determines the success or failure of 
+	// Kubernetes readiness probing.
+	ready bool
+	// started is a flag used to support readiness probing and should not be
+	// confused for affecting startup probing. When started becomes true, the
+	// proxy is done starting up.
+	started bool
+	// srv is a pointer to the HTTP server used to communicated proxy health.
+	srv *http.Server
+}
+
+// NewHealthCheck initializes a HC object and exposes HTTP endpoints used to
+// communicate proxy health.
+func NewHealthCheck(proxyClient *proxy.Client) *HC {
+	srv := &http.Server{
+		Addr: portNum,
+	}
+
+	hc := &HC{
+		live: true,
+		srv:  srv,
+	}
+
+	// Handlers used to set up HTTP endpoints.
+	http.HandleFunc(readinessPath, func(w http.ResponseWriter, _ *http.Request) {
+		readinessL.Lock()
+		hc.ready = readinessTest(proxyClient, hc)
+		if !hc.ready {
+			readinessL.Unlock()
+			w.WriteHeader(500)
+			w.Write([]byte("error\n"))
+			return
+		}
+		readinessL.Unlock()
+
+		w.WriteHeader(200)
+		w.Write([]byte("ok\n"))
+	})
+
+	http.HandleFunc(livenessPath, func(w http.ResponseWriter, _ *http.Request) {
+		livenessL.Lock()
+		hc.live = livenessTest()
+		if !hc.live {
+			livenessL.Unlock()
+			w.WriteHeader(500)
+			w.Write([]byte("error\n"))
+			return
+		}
+		livenessL.Unlock()
+
+		w.WriteHeader(200)
+		w.Write([]byte("ok\n"))
+	})
+
+	go func() {
+		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			logging.Errorf("Failed to start endpoint(s): %v", err)
+		}
+	}()
+
+	return hc
+}
+
+// CloseHealthCheck gracefully shuts down the HTTP server belonging to the HC
+// object.
+func (hc *HC) CloseHealthCheck() {
+	if hc != nil {
+		if err := hc.srv.Shutdown(context.Background()); err != nil {
+			logging.Errorf("Failed to shut down health check: ", err)
+		}
+	}
+}
+
+// NotifyReadyForConnections indicates to the proxy's HC that has finished startup.
+func (hc *HC) NotifyReadyForConnections() {
+	if hc != nil {
```

Comment:
Eno suggested that to avoid doing "if *useHealthCheck{ ... }" a second time in cloud_sql_proxy.go when calling NotifyReadyForConnections(), I should do it this way. So if hc is nil, that means the user chose not use healthchecks, so this function should not do anything.

Reply:
So this is a bit of a nuanced take, but I think it's preferable that the caller to ensure `hc` != nil. If I see a function call (that appears successful), I may assume that `hc` isn't nil, and perform additional operations for it and hit a nil pointer later on. Inside, if I see a check that `*hc != nil`, I understand that it's possible hc could be nil, and I should be careful about what I do with it.